### ISSUE TITLE
[Bug]: VBWS returns empty outputs when generation ends on a widening step

### ISSUE THREAD

**Issue by yifanQi98** [NVIDIA/TensorRT-LLM#19337]
## System Info

- CPU architecture: x86_64
- GPU: NVIDIA RTX PRO 5000 72GB Blackwell
- NVIDIA driver: 580.126.09
- TensorRT-LLM: 1.3.0rc26
- Backend: PyTorch, single GPU
- PyTorch: 2.12.0+cu130
- Transformers: 5.5.4
- Python: 3.12.7
- OS: Linux 5.10.134, glibc 2.39
- Overlap scheduler: enabled (default)

## Who can help?

@NVIDIA/trt-llm-devs

## Information

- [ ] The official example scripts
- [x] My own modified scripts

## Tasks

- [ ] An officially supported task in the `examples` folder
- [x] My own task or dataset (the problem is model- and dataset-independent after logits are produced)

## Reproduction

Variable-Beam-Width Search returns the requested number of output slots, but many outputs have empty `token_ids` when generation terminates on a widening step.

A minimal LLM API setup is:

```python
import os

from tensorrt_llm import LLM, SamplingParams

model = os.environ["MODEL_PATH"]  # Any locally available causal LM; observed with Qwen3.
llm = LLM(
    model=model,
    tokenizer=model,
    dtype="bfloat16",
    max_batch_size=1,
    max_input_len=128,
    max_seq_len=131,
    max_beam_width=100,
)

sampling_params = SamplingParams(
    n=100,
    best_of=100,
    beam_width_array=[16, 32, 100],
    use_beam_search=True,
    max_tokens=3,
    temperature=0.0,
    ignore_eos=True,
)

request_output = llm.generate(["Hello"], sampling_params)[0]
lengths = [len(output.token_ids) for output in request_output.outputs]
print(len(request_output.outputs))
print({length: lengths.count(length) for length in set(lengths)})

assert len(request_output.outputs) == 100
assert all(length == 3 for length in lengths)
```

With lightweight instrumentation around `beam_search_sampling_batch_cba`, the expected `(beam_width_in, beam_width_out)` transitions are:

```text
(1, 16), (16, 32), (32, 100)
```

With the default overlap scheduler, the observed transitions are:

```text
(1, 16), (16, 16), (16, 32)
```

The final step creates 32 beams, but `_prepare_beam_history_cba()` then collects only 16 active paths. In five repeated requests, the final output length histogram was consistently:

```text
{3: 16, 0: 84}
```

The same request with a constant schedule `[100, 100, 100]` returns 100 non-empty three-token outputs.

Set `TLLM_DEBUG_MODE=1` for additional logs.

## Expected behavior

For `beam_width_array=[16, 32, 100]` and `max_tokens=3`:

1. Sampling should execute widths `1 -> 16`, `16 -> 32`, and `32 -> 100`.
2. The final response should contain 100 outputs.
3. Every output should contain three generated tokens; no output should have an empty `token_ids` list.

## actual behavior

The overlap path repeats the first configured width for the second sampling step, so the final step widens only from 16 to 32. Finalization then slices the current-step snapshot using the input width (16), dropping the additional 16 paths. The remaining output slots stay padding and are exposed as empty `token_ids`.

## additional notes

This appears to be two issues in the Python PyExecutor VBWS path:

1. `LlmRequest.get_beam_width_by_iter()` indexes `beam_width_array` with `self.decoding_iter`. Under the overlap scheduler, the current Python sampling state is tracked by `py_decoding_iter`, while `decoding_iter` is synchronized later during response handling. This makes the width schedule lag by one step.

2. `_prepare_beam_history_cba()` computes:

   ```python
   active_width = _get_beam_width_in(request)
   ```

   However, the snapshot being finalized already contains the current step's output beams. On a widening terminal step, it should collect the produced/output width, not the input width.

A local diagnostic patch that:

- indexes the Python width schedule with `py_decoding_iter`, and
- finalizes using `request.get_beam_width_by_iter(for_next_iteration=True)`

restores the expected transitions and produces 100 non-empty three-token outputs for all tested prompts. Applying only either change produces 32 valid outputs, while applying both produces all 100.

The relevant upstream implementation was introduced/completed in #16620. Current `main` still appears to use `decoding_iter` for the Python width lookup and `_get_beam_width_in(request)` for CBA finalization.

No private model paths, prompts, endpoints, or datasets are included in this report.

## Before submitting a new issue...

- [x] I searched existing and past issues, documentation, and examples for the same problem.

**Comment by Lipenghee** [comment]
I can take this if the issue is approved for contribution. Current main still has both reported state mismatches: the Python width lookup uses `decoding_iter` while the overlap path advances `py_decoding_iter`, and CBA finalization slices the terminal snapshot with the input width.

My plan is to add focused regressions for (1) overlap-scheduled VBWS width progression and (2) finalization on a widening terminal step, then make the minimal two-site correction and run the affected beam-search unit suite plus repository pre-commit checks. I will keep fixed-width, non-overlap, C++ decoding, and unrelated beam-search behavior out of scope. The exact RTX PRO 5000 end-to-end case would remain a reporter/CI validation; the local acceptance gate will require deterministic tests that fail on current main and pass only with the fix.

Please confirm that this issue is approved for implementation under the contribution policy.

**Comment by yifanQi98** [comment]
As the reporter, I would like to claim this fix.

Both corrections are already implemented and validated locally. I re-checked the latest `main` I pulled (`7353a22eab`) and the two sites are unchanged:

- `tensorrt_llm/_torch/pyexecutor/llm_request.py` — `get_beam_width_by_iter()` indexes `beam_width_array` with `self.decoding_iter`, which lags `py_decoding_iter` while the overlap scheduler samples the next batch.
- `tensorrt_llm/_torch/pyexecutor/sampler/beam_search.py` — `_prepare_beam_history_cba()` finalizes with `active_width = _get_beam_width_in(request)`, i.e. the input width of a step whose snapshot already holds the produced beams.

The patch is minimal and scoped to those two Python PyExecutor sites: index the width schedule with the current Python decoding iteration, and finalize using the width the current step produced. There is no C++ decoding change, no API or serialized-format change, and fixed-width beam search as well as requests without `beam_width_array` keep their existing behavior.

Regression coverage I have ready, which fails on current `main` and passes only with both corrections:

- VBWS `beam_width_array=[16, 32, 100]` with `max_tokens=3`, overlap scheduler enabled and disabled;
- per-step width transitions assert `1 -> 16`, `16 -> 32`, `32 -> 100`;
- 100 returned outputs, each containing three generated tokens, no empty `token_ids`, and no `BEAM_SEARCH_PAD_TOKEN` in request histories;
- CBA finalization when a terminal step widens from 32 input beams to 100 output beams;
- fixed-width beam search as a regression guard.

Could maintainers please triage this issue and, if it is approved for external contribution, add `bug`, `Decoding/Sampling` and `Community want to contribute`? I will open the PR against `main` as soon as it is approved, and I am happy to adjust the approach based on feedback.
